modules/host/storagemanager/sector.go

```diff
@@ -238,6 +242,112 @@ func (sm *StorageManager) AddSector(sectorRoot crypto.Hash, expiryHeight types.B
 	return sm.save()
 }
 
+// AddSectorBatch will add a set of sectors to the host, correctly selecting
+// the storage folders in which the sectors belong. The sectors should all be
+// virtual sectors.
+func (sm *StorageManager) AddSectorBatch(sectorRoots []crypto.Hash, expiryHeight types.BlockHeight) error {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	err := sm.db.Update(func(tx *bolt.Tx) error {
+		var err error
+		var i int
+		var root crypto.Hash
+		for i, root = range sectorRoots {
+			// Verify that the sector is virtual.
+			sectorKey := sm.sectorID(root[:])
+			bsu := tx.Bucket(bucketSectorUsage)
+			usageBytes := bsu.Get(sectorKey)
+			var usage sectorUsage
+			if usageBytes == nil {
+				// usageBytes == nil means that this is a new sector, therefore not
+				// a virtual sector, and therefore incompatible with
+				// 'AddSectorBatch'.
+				err = errNonVirtualSectorInBatch
+				break
+			}
+			err = json.Unmarshal(usageBytes, &usage)
+			if err != nil {
+				break
+			}
+			// If the sector already has the maximum number of virtual sectors,
+			// return an error. The host handles virtual sectors differently
+			// from physical sectors and therefore needs to limit the number of
+			// times that the same data can be uploaded to the host. For
+			// renters that are properly using encryption and are using
+			// sane/reasonable file contract renewal practices, this limit will
+			// never be reached (sane behavior will cause 3-5 at an absolute
+			// maximum, but the limit is substantially higher).
+			if len(usage.Expiry) >= maximumVirtualSectors {
+				err = errMaxVirtualSectors
+				break
+			}
+			usage.Expiry = append(usage.Expiry, expiryHeight)
+			usageBytes, err = json.Marshal(usage)
+			if err != nil {
+				break
+			}
+			err = bsu.Put(sm.sectorID(root[:]), usageBytes)
+			if err != nil {
+				break
+			}
+		}
+		if err != nil {
+			// Need to remove the virtual sectors that are added, as there has
+			// been an error.
```

Comment:
oh yeah, forgot that this function entirely happens within the bolt Update call. Pretty sure there are no side-effects, because every sector is supposed to be virtual.

derp, could have saved my self some time.